ldms/src/sampler/zfs_leafvdevs/zfs_leafvdevs.c

```diff
@@ -0,0 +1,496 @@
+
+	if (nvlist_lookup_uint64_array(nvroot,
+				       ZPOOL_CONFIG_VDEV_STATS,
+				       (uint64_t **) & vs, &c) != 0) {
+		rc = 1;
+	}
+
+	record_instance = ldms_record_alloc(sampler_base->set,
+					    index_store.vdev_rec_idx);
+
+	if (record_instance == NULL) {
+		log_fn(LDMSD_LDEBUG,
+		       SAMP
+		       ": ldms_record_alloc() failed, resizing metric set\n");
+		resize_metric_set();
+		record_instance = ldms_record_alloc(sampler_base->set,
+						    index_store.vdev_rec_idx);
+		if (record_instance == NULL)
+			rc = 2;
+	}
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+					     record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[0],
+					  pool_name);
+		/* zpool state  1 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[1],
+					  zpool_state_to_name((vdev_state_t)
+							      vs->vs_state,
+							      (vdev_aux_t) vs->
+							      vs_aux));
+		/* vdevname     2 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[2],
+					  vdev_name);
+		/* vdevpath     3 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[3],
+					  vdev_path);
+		/* vdevguid     4 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[4],
+					  vdev_guid);
+		/* alloc        5 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[5],
+				    vs->vs_alloc);
+		/* free         6 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[6],
+				    vs->vs_space - vs->vs_alloc);
+		/* size         7 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[7],
+				    vs->vs_space);
+		/* read_bytes   8 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[8],
+				    vs->vs_bytes[ZIO_TYPE_READ]);
+		/* iread_errors 9 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[9],
+				    vs->vs_read_errors);
+		/* read_ops     10 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[10],
+				    vs->vs_ops[ZIO_TYPE_READ]);
+		/* write_bytes  11 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[11],
+				    vs->vs_bytes[ZIO_TYPE_WRITE]);
+		/* write_errors 12 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[12],
+				    vs->vs_write_errors);
+		/* write_ops    13 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[13],
+				    vs->vs_ops[ZIO_TYPE_WRITE]);
+		/* checksum errors 14 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[14],
+				    vs->vs_checksum_errors);
+		/* fragmentation 15 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[15],
+				    vs->vs_fragmentation);
+		/* initialization errors 16 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[16],
+				    vs->vs_initialize_errors);
+	}
+
+	free(vdev_name);
+	free(vdev_path);
+	free(vdev_guid);
+
+	return (rc);
+}
+
+/*
+ * recursive stats printer
+ */
+static int get_recursive_stats(stat_printer_f func, nvlist_t * nvroot,
+			       const char *pool_name, int descend,
```

Comment:
The `descend` arg can be removed from this function as well.

Reply:
No more descending. Gone!